### PR TITLE
GCW-3528 Handle messages for sunshine integrations

### PR DESCRIPTION
**Pull request by patrixr**
### Case

We connect a third party integration to sunshine e.g Slack, and we
answer the customer through there

When the message is picked up by GoodChat, it does not contain any
customer information, only the type which is "business"
Nor does it contain information on the staff member that sent the message

### Strategy

- Record the message as a "System" message
- If the conversation does not exist, create it without a customerId.
  When a customer replies to the conversation, the missing customerId
  will be filled

From the customer's perspective : no difference
From the staff member's perspective: some message might not have an author, and will be marked as "system" instead (we simply don't know who sent it since someone used another source to send it)

**Note**: this is an edge case, odds of it happening are close to none since we control the integrations. It will however happen often on staging since Slack is a good way to trigger messages for testing

### Other changes

- Bit of refactoring (name improvements)
- Improved factories

### Database changes

- Remove the not null constraint on `customerId` (should never have been the case in the first place)
- Add "System" to the author type enum

### CODE REVIEW

File: lib/services/conversation_service.ts [hunk 1]
```diff
@@ -2,14 +2,15 @@ import _                        from "lodash"
 import { Conversation }         from "@prisma/client";
 import db, { Unsaved }          from "../db"
 
+
 /**
- * Creates a conversation if it doesn't already exist
+ * Creates a converastion if it doesnt already exsit
```
Comment (shreyas098) [inline lib/services/conversation_service.ts]:
```suggestion
 * Creates a conversation if it doesn't already exist
```

**Review by shreyas098 — APPROVED** [review]

**Review by abulsayyad123 — APPROVED** [review]

File: lib/services/conversation_service.ts (hunk 1 above)
Comment (patrixr) [reply lib/services/conversation_service.ts]: thanks !

File: lib/services/events/index.ts [hunk 2]
```diff
@@ -53,8 +54,8 @@ export function registerWebhookHandler(type: WebhookEventType, handler: WebhookH
  * @param {WebhookEventBase} type
  * @param {WebhookPayload} payload
  */
-export async function handleWebhookEvent(event: WebhookEventBase) {
-  await each(getHandlersForType(event.type), (h : WebhookHandler) => h(event));
+export async function handleWebhookEvent(event: WebhookEventBase, cfg: GoodChatConfig) {
```
Comment (BboyStatix) [inline lib/services/events/index.ts:57]: Why the dependency injection?

File: lib/services/events/webhooks/conversation_message.ts [hunk 3]
```diff
@@ -12,49 +12,68 @@
...
+
 /**
  *
  *
  * @export
  * @param {ConversationMessageEvent} event
  * @returns {Promise<void>}
  */
-export async function onMessageCreated(event: WebhookEventBase) : Promise<void> {
+export async function onMessageCreated(event: WebhookEventBase, cfg: GoodChatConfig) : Promise<void> {
```
Comment (BboyStatix) [inline lib/services/events/webhooks/conversation_message.ts]: I don't see `cfg` being used.

File: lib/services/events/index.ts (hunk 2 above)
Comment (BboyStatix) [reply lib/services/events/index.ts:57]:
```
export async function onConversationCreated(event: WebhookEventBase) : Promise<void> {
// ...
}
```

Shouldn't the above function then take a `config` as its second argument?

Comment (patrixr) [reply lib/services/events/index.ts:57]: if it's not using the second argument there's no need to take it. typescript allows that

File: lib/services/events/webhooks/conversation_message.ts (hunk 3 above)
Comment (patrixr) [reply lib/services/events/webhooks/conversation_message.ts]: yeah It's not

**Review by steveyken — APPROVED** [review]

**Pull request merged**